Kudu.Services/Diagnostics/ProfileManager.cs

```diff
@@ -0,0 +1,314 @@
+                }
+                else if (!string.IsNullOrEmpty(error))
+                {
+                    tracer.TraceError(error);
+                    return new HttpResponseInfo(HttpStatusCode.InternalServerError, "Profiling process failed with the following error: " + error);
+                }
+
+                return new HttpResponseInfo(HttpStatusCode.OK, string.Empty);
+            }
+            catch (Exception ex)
+            {
+                tracer.TraceError(ex);
+                return new HttpResponseInfo(HttpStatusCode.InternalServerError, ex.Message);
+            }
+            finally
+            {
+                if (outputStream != null)
+                {
+                    outputStream.Dispose();
+                }
+
+                if (errorStream != null)
+                {
+                    errorStream.Dispose();
+                }
+            }
+        }
+
+        private static int GetNextProfilingSessionId()
+        {
+            // TODO: This is not a good way to track active profiling sessions, but as of VS2015 RTM, the profiling service does not provide any API to track the current active sessions.  
+            // This is planned to be fixed in VS2015 Update 1.
+            var r = new Random();
+
+            int sessionId = r.Next(1, 255);
+            List<ProfileInfo> list = _profilingList.Values.ToList<ProfileInfo>();
+
+            while (list.Exists(item => item.SessionId == sessionId))
+            {
+                sessionId = r.Next(1, 255);
+            }
+
+            return sessionId;
+        }
+
+        private static string GetString(MemoryStream stream)
+        {
+            if (stream.Length > 0)
+            {
+                return Encoding.UTF8.GetString(stream.GetBuffer(), 0, (int)stream.Length);
+            }
+
+            return String.Empty;
+        }
+
+        private static void EnsureIdleTimer()
+        {
+            lock (_lockObject)
+            {
+                if (_profilingIdleTimer == null)
+                {
+                    _profilingIdleTimer = new Timer(_ => SafeInvoke(() => OnIdleTimer()), null, _profilingTimeout, _profilingTimeout);
+                }
+            }
+        }
+
+        private static void OnIdleTimer()
+        {
+            lock (_lockObject)
+            {
+                if (_profilingList.Count > 0)
+                {
+                    // Manually stop any profiling session which has exceeded the timeout period.
+                    // TODO: VS 2015 Update 1 should have a better way to handle this.
+                    foreach (var item in _profilingList)
+                    {
+                        if (DateTime.UtcNow - item.Value.StartTime > _profilingTimeout)
+                        {
+                            Task.Run(() => StopProfileInternalAsync(item.Key, item.Value.SessionId, true));
+                        }
+                    }
+
+                    DisposeTimerIfNecessary();
```

Comment:
since this is awaited `Task.Run(() => StopProfileInternalAsync(item.Key, item.Value.SessionId, true));` There is a race condition between that and this line.